### ISSUE TITLE
Response to PermissionError

### ISSUE THREAD

**Issue by kseistrup** [michael-lazar/jetforce#27]
Currently, jetforce responds with a “An unexpected error occurred” when stumbling upon a file it cannot read (i.e., PermissionError), and while it may be technically correct, perhaps it would be more meaningful to return “51 Not Found”. That way, the gemini captain can have files that are world-unreadable (say, Makefiles or some shellscripts or notes) and that aren't meant for public consumption. The requesting party needn't know exactly why they can't get the file. Jetforce can inform the captain that somebody has requested an unreadable file so that permissions can be corrected if they were wrong, but basically I feel that it would enhance functionality if jetforce would just ignore unreable files and directories.

**Comment by michael-lazar** [comment]
👍 , I think your explanation makes sense and this is a reasonable change to make.

**Comment by kseistrup** [comment]
Perhaps I should add that the gopher server [gophernicus](https://github.com/gophernicus/gophernicus) is deploying said strategy:

> However, the security settings (which are non-changeable) are so strict that you need to keep one thing in mind. Gophernicus will only serve world-readable content. Being readable by the server process is not enough, all files and directories MUST be world-readable or they are simply hidden from all listings and denied if a client asks for them.

Apart from the security aspect, I have found the feature useful when building a gopher site: In each directory I have a `.header`, a `.footer`, a `.body` and a `Makefile`, where `.header` usually is a breadcrumb trail, and I then use the `Makefile` to assemble the `gophermap` file from the header and other parts. These parts are only readable for the admin user, but the `gophermap` file is world-readable.

**Comment by michael-lazar** [comment]
> Currently, jetforce responds with a “An unexpected error occurred” when stumbling upon a file it cannot read (i.e., PermissionError), and while it may be technically correct, perhaps it would be more meaningful to return “51 Not Found”.

I just tried to reproduce this on the latest version of jetforce but I was unable to. The code already returns a "51 Not Found" error for me. My test was:

1. ``touch /var/gemini/hello.txt``
2. ``chmod 000 /var/gemini/hello.txt``
3. ``jetforce-client gemini://localhost/hello.txt``

```
127.0.0.1 [01/Jun/2020:21:20:41 -0400] "gemini://localhost/hello.txt" 51 Not Found 14
```

I'm wondering if you were triggering a different type of permission error? The code does already attempt to check if the file is readable and catch generic operating system errors.

https://github.com/michael-lazar/jetforce/blob/9cc4a320c2254ba72f500961cb59c27edb9280cc/jetforce/app/static.py#L92-L98

**Comment by michael-lazar** [comment]
> Apart from the security aspect, I have found the feature useful when building a gopher site: In each directory I have a .header, a .footer, a .body and a Makefile, where .header usually is a breadcrumb trail, and I then use the Makefile to assemble the gophermap file from the header and other parts. These parts are only readable for the admin user, but the gophermap file is world-readable.

Couldn't you accomplish the same thing by simply not running jetforce as the root user? I think that locking down access to your files at the operating system level will always be safer than trusting a program with unrestricted access to behave correctly.

**Comment by kseistrup** [comment]
> The code already returns a "51 Not Found" error for me.

Has this always been the case (I believe I ran 0.2.3 when I posted the issue, but I should of course have said which version I were using…)? I just tried with both an unreadable directory and an unreadable file, and both are returning “51 Not Found”.

I thought this was such a clear case that I didn't report the details, which I should have. I'm sorry for wasting your time.

> Couldn't you accomplish the same thing by simply not running jetforce as the root user?

I could never dream of running server software as the root user. I have added “DynamicUser=true” and a whole lot of other stuff to the systemd unit file that controls jetforce, so it's running with a minimal set of permissions:

```sh
$ systemd-analyze security jetforce.service | tail -1
→ Overall exposure level for jetforce.service: 1.8 OK 🙂
```

I think I can shave off a little more, but that will be an exercise for another day.

When I used the term “the admin user”, I simply meant “the user administrating the whole thing”. Another confirmation of the wise saying “Abbreviations are only helpful to the abbreviator.”

**Comment by michael-lazar** [comment]
> I thought this was such a clear case that I didn't report the details, which I should have. I'm sorry for wasting your time.

No worries! Let me know if you see that unexpected error again, my hunch is that it was likely something unreleated but it would still be worth looking at if we can reproduce it.

> I could never dream of running server software as the root user. I have added “DynamicUser=true” and a whole lot of other stuff to the systemd unit file that controls jetforce, so it's running with a minimal set of permissions:

Good to hear! If you don't mind posting it, I would be curious to see your systemd unit file. I've been meaning to update the documentation to guide people to more secure default settings.

**Comment by kseistrup** [comment]
> Let me know if you see that unexpected error again

Sure thing.

> I would be curious to see your systemd unit file

I have attached jetforce.service almost verbatim. The only thing I have changed is the domain name, that I have changed to `@MYDOMAIN@`.

It is worth noting that:

1. I decided to run jetforce on IPv6 only. Changes to `jetforce.service` are necessary for IPv4:
   * `--host ::` could/should be changed to `--host 0.0.0.0`
   * The `RestrictAddressFamilies` should have `AF_INET` added
2. I have set `DynamicUser=true`, which means that the user will take its name of the service file, minus the `.service` part. So `jetforce.service` will run under user `jetforce` and `gemini.service` will run under user `gemini` and _these users needn't exist_ — they are _dynamic_. However, the certificate files shouldn't be world readable, so I have indeed created a system user called `jetforce` that is able to read the certificates:

```sh
$ sudo ls -lR /etc/jetforce
/etc/jetforce:
total 0
drwxr-x--x 2 root jetforce 52 May 27 20:02 certs

/etc/jetforce/certs:
total 8
-rw-r----- 1 root jetforce 3311 May 27 20:02 @MYDOMAIN@.crt
-rw-r----- 1 root jetforce  288 May 27 20:02 @MYDOMAIN@.key
```

So if I changed the name of the service file to, say, `gemini.service`, the certificate files would no longer be readable by the service.

----

:paperclip: [jetforce.service](https://github.com/michael-lazar/jetforce/files/4722159/jetforce.service.txt)
